File: Data Structures & Algorithms/meeting-schedule/submission-3.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def canAttendMeetings(self, intervals: List[Interval]) -> bool:
        events = []
        for i in intervals:
            events.append((i.start,1))
            events.append((i.end,-1))

        x = 0
        for e in sorted(events):
            print(e)
            x += e[1]
            if x>=2:
                return False

        return True

File: Data Structures & Algorithms/meeting-schedule-ii/submission-3.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        priority_queue<int, vector<int>, greater<int>> pq;
        sort(intervals.begin(), intervals.end(), [](auto& a, auto& b){return a.start < b.start;});
        for(auto& i: intervals){
            if(!pq.empty() && i.start >= pq.top())
                pq.pop();
            pq.push(i.end);
        }
        return pq.size();
    }
};
